### DISCUSSION TITLE
I wrote a Julia version, now 1mins C++ (was 9min when I orig. posted) vs 2m30 (single-threaded) to 15s (multi-threaded)

### DISCUSSION THREAD

**Discussion by lawless-m** [RayTracing/raytracing.github.io#987] [Show and tell]
Hi,

I bought the first two books a few years ago and having learned quite a bit of Julia in the meantime, I thought I would do a translation. There are no tricks, just a straight copy of the C++ code into Julia.

My Julia code is here [https://github.com/lawless-m/raytracing.github.io/tree/master/Julia](https://github.com/lawless-m/raytracing.github.io/tree/master/Julia)

This is just the Week One code, so just the random spheres

On my machine the C++ version 
```
$ time ./a.out > a.out.ppm
Scanlines remaining: 0
Done.

real    9m15.028s
user    9m14.819s
sys     0m0.052s
```
The Julia version, single threaded

```
$ time julia -t 40 --project=.. -L RandomScene.jl  -e "main()" > random.ppm
real    2m29.209s
user    2m27.850s
sys     0m1.088s
```

And the multi-threaded version, with each thread doing a different scanline and I am lucky enough to have a 40 core machine and it completes in a nice 15s
```
$ time julia -t 40 --project=.. -L RandomScene.jl  -e "main()" > random.ppm
real    0m15.312s
user    5m30.550s
sys     0m2.783s
```

I'm looking forward to adding the next books in due course.

**Comment by TheRayTracer** [comment]
Don't forget to add it to the list: https://github.com/RayTracing/raytracing.github.io/wiki/Implementations

**Comment by lawless-m** [comment]
I should add that it turns out I was running the C++ with poor compiler flags - I've since used -march=native and -O 3 and the C++ code runs in under a minute.

**Reply by hbhutta** [reply]
Hi you wrote this a while ago, but I have a question. Could you explain what -march=native does exactly and how it made your code faster. I am new to this so I'd like to know.

**Reply by lawless-m** [reply]
-march=native or arch=native gives the compiler the option to produce machine code that will only run on the host CPU rather than generic instructions that will run on any compatible CPU. As you may imagine, X86 has quite a range of opcodes that are CPU model specific
